service/matching/task_queue_manager.go

```diff
@@ -783,40 +805,38 @@ func (c *taskQueueManagerImpl) userDataFetchSource() (string, error) {
 func (c *taskQueueManagerImpl) fetchUserData(ctx context.Context) error {
 	ctx = c.callerInfoContext(ctx)
 
-	if !c.config.LoadUserData() {
-		// if disabled, mark ready now
-		c.SetUserDataInitialFetch(nil)
-		return nil
-	}
 	if c.managesSpecificVersionSet() {
 		// tqm for specific version set doesn't have its own user data
-		c.SetUserDataInitialFetch(nil)
+		c.SetUserDataInitialFetch(errNoUserDataOnVersionedTQM, nil)
 		return nil
 	}
-	if c.db.DbStoresUserData() {
-		// root workflow partition "owns" user data, read it from db
-		err := c.db.loadUserData(ctx)
-		c.SetUserDataInitialFetch(err)
-		return err
-	}
 
 	// otherwise fetch from parent partition
 
 	fetchSource, err := c.userDataFetchSource()
 	if err != nil {
 		if err == errMissingNormalQueueName { // nolint:goerr113
-			// pretend we have no user data
-			c.SetUserDataInitialFetch(nil)
+			// pretend we have no user data. this is a sticky queue so the only effect is that we can't
+			// kick off versioned pollers.
+			c.SetUserDataInitialFetch(nil, nil)
 		}
 		return err
 	}
 
-	firstCall := true
+	hasFetchedUserData := false
 
 	op := func(ctx context.Context) error {
+		if !c.config.LoadUserData() {
+			// if disabled, mark disabled and ready, but allow retries so that we notice if
+			// it's re-enabled
+			c.SetUserDataInitialFetch(errUserDataDisabled, nil)
+			return errUserDataDisabled
+		}
+
 		knownUserData, _, err := c.GetUserData(ctx)
 		if err != nil {
-			return err
+			// start with a non-long poll after re-enabling after disable.
+			hasFetchedUserData = false
 		}
```

Comment:
Why does this condition mean that we just re-enabled after the feature was disabled? Also, why would we want to do a non-long poll in that case?